### PR TITLE
t3 - initial search version

### PR DESCRIPTION
**Pull request by KenXiong123**
1. Create account repo, search controller, search rsp, search service and search service impl. URL parsing is handled in search service impl (e.g. findbyID or findby other user info)

### CODE REVIEW

File: src/main/java/com/lgtm/easymoney/repositories/AccountRepository.java [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package com.lgtm.easymoney.repositories;
+
+import com.lgtm.easymoney.models.Account;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+import java.util.Optional;
+
+@Repository
+public interface AccountRepository extends JpaRepository<Account, Long> {
+    Optional<Account> findByAccountName(String accountName);
```
Comment (howieraem) [inline src/main/java/com/lgtm/easymoney/repositories/AccountRepository.java:10]: Try `List<Account> findByAccountNameContains(String accountName)`

File: src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+    private final AccountRepository accountRepository;
+
+    @Autowired
+    public SearchServiceImpl(UserRepository userRepository, AccountRepository accountRepository) {
+        this.userRepository = userRepository;
+        this.accountRepository = accountRepository;
+    }
+    
+    @Override
+    public User getUserByID(Long id) {
```
Comment (rl3250) [inline src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java:27]: reuse `userService.getUserByID()`

File: src/main/java/com/lgtm/easymoney/repositories/UserRepository.java [hunk 3]
```diff
@@ -3,8 +3,12 @@
 import com.lgtm.easymoney.models.User;
 import org.springframework.data.jpa.repository.JpaRepository;
 import org.springframework.stereotype.Repository;
+import java.util.Optional;
 
 @Repository
 public interface UserRepository extends JpaRepository<User, Long> {
     boolean existsByEmail(String email);
+
+    Optional<User> findByEmail(String email);
```
Comment (howieraem) [inline src/main/java/com/lgtm/easymoney/repositories/UserRepository.java:12]: `List<User> findByEmailContains(String email)`

File: src/main/java/com/lgtm/easymoney/repositories/UserRepository.java [hunk 4]
```diff
@@ -3,8 +3,12 @@
...
+    Optional<User> findByPhone(String phone);
```
Comment (howieraem) [inline src/main/java/com/lgtm/easymoney/repositories/UserRepository.java:13]: `List<User> findByPhoneContains(String phone)`

File: src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        
+        //regex to check valid phone number
+        String phoneRegex = "^(?:(?:\\+?1\\s*(?:[.-]\\s*)?)?(?:\\(\\s*([2-9]1[02-9]|[2-9][02-8]1|[2-9][02-8][02-9])\\s*\\)|([2-9]1[02-9]|[2-9][02-8]1|[2-9][02-8][02-9]))\\s*(?:[.-]\\s*)?)?([2-9]1[02-9]|[2-9][02-9]1|[2-9][02-9]{2})\\s*(?:[.-]\\s*)?([0-9]{4})(?:\\s*(?:#|x\\.?|ext\\.?|extension)\\s*(\\d+))?$";
+        
+        //regex to check valid email pattern
+        Pattern p = Pattern.compile("^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,6}$", Pattern.CASE_INSENSITIVE);
+        java.util.regex.Matcher matcher = p.matcher(userInfo);
+
+        //Parse userInfo
+        if(matcher.find()) {
```
Comment (howieraem) [inline src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java:95]: Do the union of results of the 3 JpaRepo functions instead

File: src/main/java/com/lgtm/easymoney/payload/SearchRsp.java [hunk 6]
```diff
@@ -0,0 +1,15 @@
+package com.lgtm.easymoney.payload;
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@AllArgsConstructor
+@Data
+@NoArgsConstructor
+public class SearchRsp {
```
Comment (howieraem) [inline src/main/java/com/lgtm/easymoney/payload/SearchRsp.java:10]: I think it's also reasonable to return userType

**Review by howieraem — COMMENTED** [review]
Better to return more users

File: src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java (hunk 2 above)
Comment (rl3250) [inline src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java:27]: integrate these `getUserByXX` into userService and call api via the userService.
question is : some of the fields are not unique, maybe email? how to handle this properly? potentially a list of users....

the search service should be robust imho, thus i think we need to refactor the response and handle 1) one client 2) no result 3) multiple users match

**Review by rl3250 — CHANGES_REQUESTED** [review]
back to your queue to address comments above

File: src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java (hunk 2 above)
Comment (KenXiong123) [reply src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java:27]: Emails are unique. The same email cannot be registered twice. 

For the integration of these methods, I will gauge other team members' interests. Please refer to the slack channel.

File: src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+        
+        //regex to check valid phone number
+        String phoneRegex = "^(?:(?:\\+?1\\s*(?:[.-]\\s*)?)?(?:\\(\\s*([2-9]1[02-9]|[2-9][02-8]1|[2-9][02-8][02-9])\\s*\\)|([2-9]1[02-9]|[2-9][02-8]1|[2-9][02-8][02-9]))\\s*(?:[.-]\\s*)?)?([2-9]1[02-9]|[2-9][02-9]1|[2-9][02-9]{2})\\s*(?:[.-]\\s*)?([0-9]{4})(?:\\s*(?:#|x\\.?|ext\\.?|extension)\\s*(\\d+))?$";
+        
+        //regex to check valid email pattern
+        Pattern p = Pattern.compile("^[A-Z0-9._%+-]+@[A-Z0-9.-]+\\.[A-Z]{2,6}$", Pattern.CASE_INSENSITIVE);
+        java.util.regex.Matcher matcher = p.matcher(userInfo);
+
+        //Parse userInfo
+        if(matcher.find() || userInfo.matches(phoneRegex)) {
```
Comment (howieraem) [inline src/main/java/com/lgtm/easymoney/services/impl/SearchServiceImpl.java:105]: What if the client wants list of accountname matched union list of email matched?

**Review by howieraem — COMMENTED** [review]
Probably no need to use regex

**Review by howieraem — APPROVED** [review]

**Pull request merged**